test(endpoints_tests): add tests for app

AlexMomot-717 opened the pull request.

Commit:
test(endpoints_tests): add tests for app

svntmr commented on .github/workflows/hexlet-check.yml:
should be moved to a separate commit with an explanation of why we do that

svntmr commented on .pre-commit-config.yaml:
should be included in the same commit as the one where `hexlet-check.yml` is deleted

svntmr commented on README.md:
should be included in the same commit as the one where `hexlet-check.yml` is deleted

svntmr commented on tests/conftest.py:
```diff
@@ -1,9 +1,27 @@
+
+@pytest.fixture
+def client(app):
+    return app.test_client()
```
as we're not doing any complicated setup in the `app` fixture, we can unify them into one. wdyt?

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,235 @@
+    url_id = get_url_by_name(url_name)
+    assert url_id == 1
+    assert response.request.path == "/urls/1"
+    assert "Страница успешно добавлена" in response.text
```
it's better to do response checks first and then check the side effects (DB entry creation)

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,235 @@
+)
+
+
+def test_index(client):
```
please, add type hints

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,235 @@
+
+    # then
+    assert response.status_code == 200
+    assert "Бесплатно проверяйте сайты на SEO-пригодность" in response.text
```
wdyt, is it enough to know that we return the stable and reproducible page content?

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,235 @@
+
+
+@freeze_time("2024-05-01")
+def test_show_url_checks(test_database, client):
```
```suggestion
from flask.testing import FlaskClient

def test_show_url_with_checks(test_database: None, client: FlaskClient) -> None:
```

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,235 @@
+    expected_text_elements = ["1", "2", url_name1, url_name2, "200", "2024-05-01"]
+    assert response.status_code == 200
+    for text in expected_text_elements:
+        assert text in response.text
```
too complicated to understand what's checked

Commit:
chore: remove hexlet checks

since there will be no changes to the API and this pipeline is constantly failing due to having its own dependencies that miss freezegun library

svntmr commented on tests/conftest.py:
```diff
@@ -1,9 +1,21 @@
 
 
+@pytest.fixture
+def client():
```
let's add type hints!

svntmr commented on tests/fixtures/expected_htmls.py:
wdyt, does it make sense to move values to actual HTML files? and load their content here

AlexMomot-717 replied:
it will work for clear index.html and page_not_found.html only, cause those don't depend on context from endpoints

svntmr replied:
what context?

AlexMomot-717 replied:
I mean (context is about django :)) url_data, checks_data etc

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,239 @@
+
+    # then
+    assert response.status_code == 200
+    assert EXPECTED_INDEX_HTML in response.text
```
why not `==`?

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,239 @@
+    response = client.post(route, data={"url": url_name}, follow_redirects=True)
+
+    # then
+    assert "Страница уже существует" in response.text
```
let's add html for this page

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,239 @@
+
+    # then
+    assert response.status_code == 200
+    assert "Страница не найдена" in response.text
```
let's add html content here

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,239 @@
+            "created_at": datetime.date(2024, 5, 1),
+        }
+    ]
+    assert response.status_code == 200
```
let's check response first and side effect later :)

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,239 @@
+
+    # then
+    assert response.status_code == 200
+    assert "Страница не найдена" in response.text
```
let's add html here

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,249 @@
+
+def expected_html_data(html_file):
+    with open(html_file, "r", encoding="utf-8") as f:
+        return f.read()
```
you can user `resolve_path` and `Path.read_bytes()` instead

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,249 @@
+    # then
+    html_file = str(resolve_path("fixtures/expected_index.html"))
+    assert expected_html_data(html_file) == response.text
+    assert response.status_code == 200
```
```suggestion
    expected_html = resolve_path("fixtures/expected_index.html")
    assert response.text == expected_html.read_bytes()
    assert response.status_code == 200
```

svntmr reviewed (changes requested):
close!

svntmr commented on .pre-commit-config.yaml:
```diff
@@ -45,14 +47,15 @@ repos:
     rev: 7.0.0
     hooks:
       - id: flake8
+        exclude: tests/fixtures
```
I guess we can remove this exclude now as fixtures are HTML files now - flake shouldn't care about them

svntmr commented on page_analyzer/utils/db.py:
```diff
@@ -89,7 +89,9 @@ def get_urls() -> List[Dict[str, Any]]:
 
 
-def create_check(url_data: Dict[str, Any]) -> bool:
+def create_check(url_data: Dict[str, Union[int, str]] | None) -> bool:
```
In modern versions of Python ([since 3.10](https://docs.python.org/3/library/typing.html#typing.Union)) `Union` can be replaced with `|`

```suggestion
def create_check(url_data: Dict[str, int | str] | None) -> bool:
```

svntmr commented on page_analyzer/utils/db.py:
```diff
@@ -89,7 +89,9 @@ def get_urls() -> List[Dict[str, Any]]:
-def create_check(url_data: Dict[str, Any]) -> bool:
+def create_check(url_data: Dict[str, Union[int, str]] | None) -> bool:
+    if url_data is None:
+        return False
```
will this function ever be called with `None` as input?

AlexMomot-717 replied:
you're right:  
get_url_data(url_id) can return  None,
but function add_url_check catches that None:

def add_url_check(id: int) -> Any:
    url_data = get_url_data(id)
    if not url_data:
        return render_template("page_not_found.html")
    check_context = create_check(url_data)

However mypy ignores this fact and requires optional None

svntmr replied:
mypy won't know about 'None' if you remove it from the arguments typing

svntmr commented on pyproject.toml:
```diff
@@ -50,10 +50,12 @@ line_length = 79
 strict = true
+ignore_missing_imports = true
+disallow_untyped_decorators = false
+strict_equality = false
```
what was the reason for adding these settings?

AlexMomot-717 replied:
![image](https://github.com/AlexMomot-717/python-project-83/assets/76219029/6ab71d14-1cea-485c-8164-d1e4e2fdb0b2)

AlexMomot-717 replied:
tests/conftest.py:21: error: Untyped decorator makes function "test_database" untyped

tests/utils/test_db.py:25: error: Non-overlapping equality check (left operand type: "Optional[Dict[str, Union[int, str]]]", right operand type: "List[object]")
tests/utils/test_db.py:39: error: Non-overlapping equality check (left operand type: "Optional[Dict[str, Union[int, str]]]", right operand type: "List[object]")
tests/utils/test_db.py:88: error: Non-overlapping equality check (left operand type: "List[Dict[str, Any]]", right operand type: "List[List[object]]")

svntmr replied:
first issues can be resolved by a couple of isinstance asserts

svntmr commented on tests/conftest.py:
```diff
@@ -1,11 +1,25 @@
+
+
+@pytest.fixture
+def test_database() -> Generator[Any, None, None]:
```
as `Generator` is typed using `Generator[YieldType, SendType, ReturnType]` ([context](https://docs.python.org/3/library/typing.html#typing.Generator)), this function should be defined in this way:
```suggestion
def test_database() -> Generator[None, None, None]:
```

svntmr commented on tests/fixtures/__init__.py:
this folder is not a python package, this file can be removed

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,242 @@
+    expected_html = resolve_path("fixtures/expected_url_add.html")
+    assert response.data == expected_html.read_bytes()
+    url_id = get_url_by_name(url_name)
+    assert url_id == 1
```
[optional] can be inlined

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,242 @@
+
+    # then
+    expected_html = resolve_path("fixtures/expected_url_exists.html")
+    assert response.data == expected_html.read_bytes()
```
what is the response code in this case?

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,242 @@
+
+    # then
+    expected_html = resolve_path("fixtures/expected_url_add.html")
+    assert response.data == expected_html.read_bytes()
```
what's the response code?

svntmr commented on tests/test_app.py:
```diff
@@ -0,0 +1,242 @@
+    route = f"/urls/{url_id}/checks"
+
+    # when
+    with patch("page_analyzer.utils.db.check_url", return_value=None):
+        response = client.post(route, follow_redirects=True)
```
should this part raise instead? test function says `error`

AlexMomot-717 replied:
I consider this side effect could be raised inside check_url test.
we get None here as a result of check_url raise error.

svntmr replied:
let's rename this test to smth like "...check_fails"

svntmr replied:
@AlexMomot-717 bump

svntmr reviewed (approved):
one small comment from the previous review
